frame/pallet-account-alias-registry/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+			let name =
+				sp_std::str::from_utf8(&name[..]).map_err(|_| Error::<T>::InvalidNameFormat)?;
+			let tag =
+				T::TagProvider::tag(&who, name).map_err(|_| Error::<T>::FailedTagGeneration)?;
+			let account_name =
+				AccountName::new(&name, tag).map_err(|_| Error::<T>::InvalidNameFormat)?;
+			AccountAliases::<T>::try_mutate(
+				AccountAlias::AccountName(account_name),
+				|maybe_value| -> DispatchResult {
+					ensure!(maybe_value.is_none(), Error::<T>::InUse);
+					*maybe_value = Some(who.clone());
+					Ok(())
+				},
+			)?;
+			AccountNamesIndex::<T>::insert(&who, account_name);
+			Self::deposit_event(Event::<T>::AccountNameAliased { who, aliased: account_name });
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::reclaim_account_name())]
+		pub fn reclaim_account_name(origin: OriginFor<T>, new_name: Vec<u8>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let account_name = AccountNamesIndex::<T>::get(&who).ok_or(Error::<T>::NotAliased)?;
+			let new_name =
+				sp_std::str::from_utf8(&new_name[..]).map_err(|_| Error::<T>::InvalidNameFormat)?;
+			let tag =
+				T::TagProvider::tag(&who, new_name).map_err(|_| Error::<T>::FailedTagGeneration)?;
+			let new_account_name =
+				AccountName::new(&new_name, tag).map_err(|_| Error::<T>::InvalidNameFormat)?;
+			AccountAliases::<T>::try_mutate(
+				AccountAlias::AccountName(new_account_name),
+				|maybe_value| -> DispatchResult {
+					ensure!(maybe_value.is_none(), Error::<T>::InUse);
+					*maybe_value = Some(who.clone());
+					Ok(())
+				},
+			)?;
+			AccountAliases::<T>::remove(AccountAlias::AccountName(account_name));
+			AccountNamesIndex::<T>::insert(&who, new_account_name);
+			Self::deposit_event(Event::<T>::AccountNameRealiased {
+				who,
+				past_aliased: account_name,
+				current_aliased: new_account_name,
+			});
+			Ok(())
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::claim_k1_address())]
+		pub fn claim_k1_address(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let ethereum_address = Self::claim_k1_address_from_id(&who)?;
+			Self::deposit_event(Event::<T>::EthAddressAliased { who, aliased: ethereum_address });
+			Ok(())
+		}
+
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::force_claim_account_name())]
+		pub fn force_claim_account_name(
+			origin: OriginFor<T>,
+			dest: AccountIdLookupOf<T>,
+			name: Vec<u8>,
+			tag: u16,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let dest = T::Lookup::lookup(dest)?;
+			ensure!(tag < 10000, Error::<T>::NotAllowedTag);
+			let name =
+				sp_std::str::from_utf8(&name[..]).map_err(|_| Error::<T>::InvalidNameFormat)?;
+			let new_account_name =
+				AccountName::new(&name, tag).map_err(|_| Error::<T>::InvalidNameFormat)?;
+			if let Some(past_name) = AccountNamesIndex::<T>::get(&dest) {
+				AccountAliases::<T>::remove(AccountAlias::AccountName(past_name));
+			}
```

Comment:
You should consider put this removal after account name assignment. Account name assignment can fail if the same name exists.